larray_editor/arrayadapter.py

```diff
@@ -3,188 +3,557 @@
-        # test self.filtered_data.size == 0 is required in case an instance built as LArray([]) is passed
-        # test len(axes) == 0 is required when a user filters until to get a scalar
-        if self.filtered_data.size == 0 or len(axes) == 0:
-            return [[]]
-        else:
-            axes_names = axes.display_names
-            if len(axes_names) >= 2:
-                axes_names = axes_names[:-2] + [axes_names[-2] + '\\' + axes_names[-1]]
-            return [[axis_name] for axis_name in axes_names]
-
-    def get_xlabels(self):
-        axes = self.filtered_data.axes
-        if self.filtered_data.size == 0 or len(axes) == 0:
-            return [[]]
+    @property
+    def data(self):
+        return self._original_data
+
+    @data.setter
+    def data(self, original_data):
+        assert original_data is not None, "{} does not accept None as input data".format(self.__class__)
+        self._original_data = self.prepare_data(original_data)
+
+    @property
+    def bg_value(self):
+        return self._bg_value
+
+    @bg_value.setter
+    def bg_value(self, bg_value):
+        self._bg_value = self.prepare_bg_value(bg_value)
+
+    @property
+    def changes(self):
+        return self._changes
+
+    @changes.setter
+    def changes(self, changes):
+        if changes is None:
+            self._changes = {}
         else:
-            # this is a lazy equivalent of:
-            # return [(label,) for label in axes.labels[-1]]
-            return Product([axes.labels[-1]])
-
-    def get_ylabels(self):
-        axes = self.filtered_data.axes
-        if self.filtered_data.size == 0 or len(axes) == 0:
-            return [[]]
+            assert isinstance(changes, dict), "{} only accept None or a dict as input changes".format(self.__class__)
+            self._changes = changes
+
+    # ===================== #
+    #  METHODS TO OVERRIDE  #
+    # ===================== #
+
+    def prepare_data(self, data):
+        """Must be overridden if data passed to set_data need some checks and/or transformations"""
+        return data
+
+    def prepare_bg_value(self, bg_value):
+        """Must be overridden if bg_value passed to set_data need some checks and/or transformations"""
+        return bg_value
+
+    def filter_data(self, data, filter):
+        """Return filtered data"""
+        raise NotImplementedError()
+
+    def get_axes(self, data):
+        """Return list of :py:class:`Axis` or an empty list in case of a scalar or an empty array.
+        """
+        raise NotImplementedError()
+
+    def _get_raw_data(self, data):
+        """Return internal data as a ND Numpy array"""
+        raise NotImplementedError()
+
+    def _get_bg_value(self, bg_value):
+        """Return bg_value as ND Numpy array or None.
+        It must have the same shape as data if not None.
+        """
+        raise NotImplementedError()
+
+    def from_selection(self, raw_data, axes_names, vlabels, hlabels):
```

Comment:
I think this method should be "protected" (with a leading _) because it should not be called from outside the adapter.

Reply:
I wanted to add an action `extract` in the widget (in addition to copy, copy to Excel, paste and plot). This will popup a box dialog asking for the name of the new variable containing the subset.

Reply:
Ok. It seems like a good idea in the long run, but I think it is premature as we need MultiIndex/sparse for this (or we will have to "densify" the result which would be confusing).